gtsfm/pose_slam/pose_slam.py

```diff
@@ -20,6 +23,106 @@
 
 
 class PoseSlam:
+    """Initializes using PoseSLAM given relative pose priors."""
+
+    @staticmethod
+    def angle(R1, R2):
+        """Calculate angle between two rotations, in degrees."""
+        return np.degrees(np.linalg.norm(R1.logmap(R2)))
+
+    @staticmethod
+    def difference(P1, P2):
+        """Calculate the translation and angle differences of two poses.
+        P1, P2: Pose3
+        Return:
+            distance: translation difference
+            angle: angular difference
+        """
+        # TODO(frank): clean this up
+        t1 = P1.translation()
+        t2 = P2.translation()
+        R1 = P1.rotation()
+        R2 = P2.rotation()
+        R1_2 = R1.compose(R2.inverse())
+        t1_ = R1_2.rotate(t2)
+        # t1_2 = t1 - R1_2*t2
+        distance = np.linalg.norm(t1 - t1_)
+        angle_ = PoseSlam.angle(R1, R2)
+        return distance, angle_
+
+    @staticmethod
+    def check_covariance(cov: np.ndarray) -> bool:
+        """Return false if covariance is bad."""
+        if np.isnan(cov).any():
+            return False
+        try:
+            info = np.linalg.inv(cov)
+            if np.isnan(info).any():
+                return False
+            else:
+                return True
+        except np.linalg.LinAlgError:
+            return False
+
+    @staticmethod
+    def filter_constraints(constraints: List[Constraint], poses: List[Pose3]) -> List[Constraint]:
+        """Filter constraints according to notebook criteria.
+
+        Args:
+            constraints (List[Constraint]): relative constraints from file
+            poses (List[Pose3]): estimated initial values
+
+        Returns:
+            Filtered constraints.
+        """
+
+        filtered_constraints: List[Constraint] = []
+
+        # Filter on covariance and error
+        for constraint in constraints:
+            a, b = constraint.a, constraint.b
+            aTb, cov = constraint.aTb, constraint.cov
+            predicted_aTb = poses[a].between(poses[b])
+            trans_diff, rot_diff = PoseSlam.difference(aTb, predicted_aTb)
+            inlier = (trans_diff <= 0.04) and (rot_diff <= 5)
+            if inlier and PoseSlam.check_covariance(cov):
+                filtered_constraints.append(constraint)
+
+        return filtered_constraints
+
+    @staticmethod
+    def filtered_pose_priors(
```

Comment:
@dellaert where is this function used? Or intended to be used?

If we plan to replace the call to HiltiLoader's get_relative_pose_priors(), we should do the following:
1. transform the prior to CAM2-CAM2
2. Add the *star* configuration hard priors on the sub-sampled timestamps.

Reply:
This is filtering the constraints that come from the constraints file. So, it comes *before* that function call, at loading.